Cannot delete a text exercise / text submission when text blocks exist

### Describe the bug

When deleting a course with a text exercise or a text exercise or a text submission with text blocks, the following exception occurs on the server:

```
Jan 08 13:10:46 vmbruegge105 java[10318]: 2020-01-08 13:10:46.522 ERROR 10318 --- [nio-8080-exec-2] o.h.engine.jdbc.spi.SqlExceptionHelper   : Cannot delete or update a parent row: a foreign key constraint fails (`ArTEMiS`.`text_block`, CONSTRAINT `fk_text_block_submission_id` FOREIGN KEY (`submission_id`) REFERENCES `submission` (`id`) ON DELETE RESTRICT ON UPDATE RESTRICT)
Jan 08 13:10:46 vmbruegge105 java[10318]: 2020-01-08 13:10:46.552 ERROR 10318 --- [nio-8080-exec-2] o.z.problem.spring.common.AdviceTraits   : Internal Server Error
Jan 08 13:10:46 vmbruegge105 java[10318]: org.springframework.dao.DataIntegrityViolationException: could not execute statement; SQL [n/a]; constraint [null]; nested exception is org.hibernate.exception.ConstraintViolationException: could not execute statement
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:297)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:254)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.translateExceptionIfPossible(AbstractEntityManagerFactoryBean.java:528)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.dao.support.ChainedPersistenceExceptionTranslator.translateExceptionIfPossible(ChainedPersistenceExceptionTranslator.java:61)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.dao.support.DataAccessUtils.translateIfNecessary(DataAccessUtils.java:242)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:153)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.data.jpa.repository.support.CrudMethodMetadataPostProcessor$CrudMethodMetadataPopulatingMethodInterceptor.invoke(CrudMethodMetadataPostProcessor.java:144)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.data.jpa.repository.support.CrudMethodMetadataPostProcessor$ExposeRepositoryInvocationInterceptor.invoke(CrudMethodMetadataPostProcessor.java:364)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:93)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at com.sun.proxy.$Proxy274.findWithEagerSubmissionsAndResultsById(Unknown Source)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at de.tum.in.www1.artemis.service.ParticipationService.delete(ParticipationService.java:773)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at de.tum.in.www1.artemis.service.ParticipationService.deleteAllByExerciseId(ParticipationService.java:864)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at de.tum.in.www1.artemis.service.ParticipationService$$FastClassBySpringCGLIB$$f26b3f60.invoke(<generated>)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:769)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:295)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:98)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:689)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at de.tum.in.www1.artemis.service.ParticipationService$$EnhancerBySpringCGLIB$$e505eaca.deleteAllByExerciseId(<generated>)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at de.tum.in.www1.artemis.service.ExerciseService.delete(ExerciseService.java:254)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at de.tum.in.www1.artemis.service.ExerciseService$$FastClassBySpringCGLIB$$376511.invoke(<generated>)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:769)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:295)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:98)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:689)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at de.tum.in.www1.artemis.service.ExerciseService$$EnhancerBySpringCGLIB$$1732f27d.delete(<generated>)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at de.tum.in.www1.artemis.web.rest.CourseResource.deleteCourse(CourseResource.java:600)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at de.tum.in.www1.artemis.web.rest.CourseResource$$FastClassBySpringCGLIB$$ea83f9d6.invoke(<generated>)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:769)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.access.intercept.aopalliance.MethodSecurityInterceptor.invoke(MethodSecurityInterceptor.java:69)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:689)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at de.tum.in.www1.artemis.web.rest.CourseResource$$EnhancerBySpringCGLIB$$c680a812.deleteCourse(<generated>)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at java.base/java.lang.reflect.Method.invoke(Method.java:567)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:190)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:138)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:105)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:893)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:798)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1040)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:943)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.servlet.FrameworkServlet.doDelete(FrameworkServlet.java:931)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at javax.servlet.http.HttpServlet.service(HttpServlet.java:666)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:113)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.boot.actuate.web.trace.servlet.HttpTraceFilter.doFilterInternal(HttpTraceFilter.java:88)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:320)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:127)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:119)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:137)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:170)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at de.tum.in.www1.artemis.security.jwt.JWTFilter.doFilter(JWTFilter.java:39)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.filter.CorsFilter.doFilterInternal(CorsFilter.java:97)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:116)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.csrf.CsrfFilter.doFilterInternal(CsrfFilter.java:124)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:74)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:56)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:215)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:178)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:358)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:271)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:94)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.filterAndRecordMetrics(WebMvcMetricsFilter.java:114)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.doFilterInternal(WebMvcMetricsFilter.java:104)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:526)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.valves.RemoteIpValve.invoke(RemoteIpValve.java:747)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:367)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:860)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1591)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at java.base/java.lang.Thread.run(Thread.java:835)
Jan 08 13:10:46 vmbruegge105 java[10318]: Caused by: org.hibernate.exception.ConstraintViolationException: could not execute statement
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.hibernate.exception.internal.SQLExceptionTypeDelegate.convert(SQLExceptionTypeDelegate.java:59)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:42)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:178)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.hibernate.engine.jdbc.batch.internal.NonBatchingBatch.addToBatch(NonBatchingBatch.java:45)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.hibernate.persister.entity.AbstractEntityPersister.delete(AbstractEntityPersister.java:3498)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.hibernate.persister.entity.AbstractEntityPersister.delete(AbstractEntityPersister.java:3755)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.hibernate.action.internal.EntityDeleteAction.execute(EntityDeleteAction.java:99)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:604)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:478)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:356)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.hibernate.event.internal.DefaultAutoFlushEventListener.onAutoFlush(DefaultAutoFlushEventListener.java:53)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.hibernate.internal.SessionImpl.autoFlushIfRequired(SessionImpl.java:1433)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1519)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.hibernate.query.internal.AbstractProducedQuery.doList(AbstractProducedQuery.java:1538)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1506)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.hibernate.query.internal.AbstractProducedQuery.getSingleResult(AbstractProducedQuery.java:1554)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.hibernate.query.criteria.internal.compile.CriteriaQueryTypeQueryAdapter.getSingleResult(CriteriaQueryTypeQueryAdapter.java:109)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.data.jpa.repository.query.JpaQueryExecution$SingleEntityExecution.doExecute(JpaQueryExecution.java:221)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.data.jpa.repository.query.JpaQueryExecution.execute(JpaQueryExecution.java:91)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.data.jpa.repository.query.AbstractJpaQuery.doExecute(AbstractJpaQuery.java:136)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.data.jpa.repository.query.AbstractJpaQuery.execute(AbstractJpaQuery.java:125)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:618)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:605)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:80)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:295)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:98)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:139)
Jan 08 13:10:46 vmbruegge105 java[10318]:         ... 150 common frames omitted
Jan 08 13:10:46 vmbruegge105 java[10318]: Caused by: java.sql.SQLIntegrityConstraintViolationException: Cannot delete or update a parent row: a foreign key constraint fails (`ArTEMiS`.`text_block`, CONSTRAINT `fk_text_block_submission_id` FOREIGN KEY (`submission_id`) REFERENCES `submission` (`id`) ON DELETE RESTRICT ON UPDATE RESTRICT)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:117)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:97)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at com.mysql.cj.jdbc.ServerPreparedStatement.serverExecute(ServerPreparedStatement.java:637)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at com.mysql.cj.jdbc.ServerPreparedStatement.executeInternal(ServerPreparedStatement.java:418)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdateInternal(ClientPreparedStatement.java:1092)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdateInternal(ClientPreparedStatement.java:1040)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at com.mysql.cj.jdbc.ClientPreparedStatement.executeLargeUpdate(ClientPreparedStatement.java:1347)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdate(ClientPreparedStatement.java:1025)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java)
Jan 08 13:10:46 vmbruegge105 java[10318]:         at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:175)
Jan 08 13:10:46 vmbruegge105 java[10318]:         ... 177 common frames omitted
```


#### To Reproduce
<!-- Steps to reproduce the behavior: -->
1. Create a text exercise
2. Add a submission and result and make sure there are text blocks in the database
3. Delete the text exercise
4. See error

#### Expected behavior
It should always be possible to delete text exercises

- [ ] Please write a server integration test case that initially fails
- [ ] Fix the problem by properly deleting text blocks before deleting the submission that is referenced
- [ ] Make sure the newly introduced test now passes

--- a/src/main/java/de/tum/in/www1/artemis/domain/TextSubmission.java
+++ b/src/main/java/de/tum/in/www1/artemis/domain/TextSubmission.java
@@ -22,7 +22,7 @@ public class TextSubmission extends Submission implements Serializable {
     @Lob
     private String text;
 
-    @OneToMany(mappedBy = "submission")
+    @OneToMany(mappedBy = "submission", cascade = CascadeType.REMOVE)
     @JsonIgnoreProperties("submission")
     private List<TextBlock> blocks = new ArrayList<>();
 

--- a/src/test/java/de/tum/in/www1/artemis/TextExerciseIntegrationTest.java
+++ b/src/test/java/de/tum/in/www1/artemis/TextExerciseIntegrationTest.java
@@ -2,35 +2,35 @@ package de.tum.in.www1.artemis;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
+import java.util.List;
 import java.util.Optional;
 
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
 import org.springframework.security.test.context.support.WithMockUser;
 
+import de.tum.in.www1.artemis.domain.Course;
+import de.tum.in.www1.artemis.domain.TextBlock;
 import de.tum.in.www1.artemis.domain.TextExercise;
 import de.tum.in.www1.artemis.domain.TextSubmission;
 import de.tum.in.www1.artemis.domain.enumeration.Language;
 import de.tum.in.www1.artemis.domain.participation.Participation;
-import de.tum.in.www1.artemis.repository.CourseRepository;
 import de.tum.in.www1.artemis.repository.TextExerciseRepository;
 import de.tum.in.www1.artemis.repository.TextSubmissionRepository;
 import de.tum.in.www1.artemis.util.DatabaseUtilService;
 import de.tum.in.www1.artemis.util.ModelFactory;
 import de.tum.in.www1.artemis.util.RequestUtilService;
 
-public class TextExerciseTest extends AbstractSpringIntegrationTest {
+public class TextExerciseIntegrationTest extends AbstractSpringIntegrationTest {
 
     @Autowired
     DatabaseUtilService database;
 
     @Autowired
     RequestUtilService request;
-
-    @Autowired
-    CourseRepository courseRepo;
 
     @Autowired
     TextExerciseRepository textExerciseRepository;
@@ -51,16 +51,29 @@ public class TextExerciseTest extends AbstractSpringIntegrationTest {
     @Test
     @WithMockUser(value = "tutor1", roles = "TA")
     public void submitEnglishTextExercise() throws Exception {
-        database.addCourseWithOneTextExercise();
+        final Course course = database.addCourseWithOneTextExercise();
         TextSubmission textSubmission = ModelFactory.generateTextSubmission("This Submission is written in English", Language.ENGLISH, false);
-        long courseID = courseRepo.findAllActive().get(0).getId();
-        TextExercise textExercise = textExerciseRepository.findByCourseId(courseID).get(0);
-        request.postWithResponseBody("/api/courses/" + courseID + "/exercises/" + textExercise.getId() + "/participations", null, Participation.class);
+        TextExercise textExercise = textExerciseRepository.findByCourseId(course.getId()).get(0);
+        request.postWithResponseBody("/api/courses/" + course.getId() + "/exercises/" + textExercise.getId() + "/participations", null, Participation.class);
         textSubmission = request.postWithResponseBody("/api/exercises/" + textExercise.getId() + "/text-submissions", textSubmission, TextSubmission.class);
 
         Optional<TextSubmission> result = textSubmissionRepository.findById(textSubmission.getId());
         assertThat(result.isPresent()).isEqualTo(true);
         result.ifPresent(submission -> assertThat(submission.getLanguage()).isEqualTo(Language.ENGLISH));
 
+    }
+
+    @Test
+    @WithMockUser(value = "instructor1", roles = "INSTRUCTOR")
+    public void deleteTextExerciseWithSubmissionWithTextBlocks() throws Exception {
+        final Course course = database.addCourseWithOneTextExercise();
+        TextExercise textExercise = textExerciseRepository.findByCourseId(course.getId()).get(0);
+        TextSubmission textSubmission = ModelFactory.generateTextSubmission("Lorem Ipsum Foo Bar", Language.ENGLISH, true);
+        textSubmission = database.addTextSubmission(textExercise, textSubmission, "student1");
+
+        final List<TextBlock> textBlocks = List.of(ModelFactory.generateTextBlock(0, 11, "Lorem Ipsum"), ModelFactory.generateTextBlock(12, 19, "Foo Bar"));
+        database.addTextBlocksToTextSubmission(textBlocks, textSubmission);
+
+        request.delete("/api/text-exercises/" + textExercise.getId(), HttpStatus.OK);
     }
 }

--- a/src/test/java/de/tum/in/www1/artemis/TextSubmissionIntegrationTest.java
+++ b/src/test/java/de/tum/in/www1/artemis/TextSubmissionIntegrationTest.java
@@ -15,6 +15,7 @@ import org.springframework.http.HttpStatus;
 import org.springframework.security.test.context.support.WithMockUser;
 
 import de.tum.in.www1.artemis.domain.Exercise;
+import de.tum.in.www1.artemis.domain.TextBlock;
 import de.tum.in.www1.artemis.domain.TextExercise;
 import de.tum.in.www1.artemis.domain.TextSubmission;
 import de.tum.in.www1.artemis.domain.User;
@@ -169,5 +170,15 @@ public class TextSubmissionIntegrationTest extends AbstractSpringIntegrationTest
         participationRepository.save(afterDueDateParticipation);
 
         request.put("/api/exercises/" + textExerciseBeforeDueDate.getId() + "/text-submissions", textSubmission, HttpStatus.OK);
+    }
+
+    @Test
+    @WithMockUser(value = "instructor1", roles = "INSTRUCTOR")
+    public void deleteTextSubmissionWithTextBlocks() throws Exception {
+        textSubmission = database.addTextSubmission(textExerciseBeforeDueDate, textSubmission, "student1");
+        final List<TextBlock> blocks = List.of(ModelFactory.generateTextBlock(0, 9), ModelFactory.generateTextBlock(10, 19), ModelFactory.generateTextBlock(20, 29));
+        database.addTextBlocksToTextSubmission(blocks, textSubmission);
+
+        request.delete("/api/submissions/" + textSubmission.getId(), HttpStatus.OK);
     }
 }

--- a/src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java
+++ b/src/test/java/de/tum/in/www1/artemis/util/DatabaseUtilService.java
@@ -84,6 +84,9 @@ public class DatabaseUtilService {
     TextSubmissionRepository textSubmissionRepo;
 
     @Autowired
+    TextBlockRepository textBlockRepo;
+
+    @Autowired
     FileUploadSubmissionRepository fileUploadSubmissionRepo;
 
     @Autowired
@@ -830,6 +833,14 @@ public class DatabaseUtilService {
         submission = textSubmissionRepo.save(submission);
         result = resultRepo.save(result);
         studentParticipationRepo.save(participation);
+        return submission;
+    }
+
+    public TextSubmission addTextBlocksToTextSubmission(List<TextBlock> blocks, TextSubmission submission) {
+        blocks.forEach(block -> block.setSubmission(submission));
+        submission.setBlocks(blocks);
+        textBlockRepo.saveAll(blocks);
+        textSubmissionRepo.save(submission);
         return submission;
     }
 

--- a/src/test/java/de/tum/in/www1/artemis/util/ModelFactory.java
+++ b/src/test/java/de/tum/in/www1/artemis/util/ModelFactory.java
@@ -267,4 +267,17 @@ public class ModelFactory {
         result.setScore(score);
         return result;
     }
+
+    public static TextBlock generateTextBlock(int startIndex, int endIndex, String text) {
+        final TextBlock textBlock = new TextBlock();
+        textBlock.setStartIndex(startIndex);
+        textBlock.setEndIndex(endIndex);
+        textBlock.setText(text);
+        textBlock.computeId();
+        return textBlock;
+    }
+
+    public static TextBlock generateTextBlock(int startIndex, int endIndex) {
+        return generateTextBlock(startIndex, endIndex, "");
+    }
 }

